R/input-submit.R

```diff
@@ -0,0 +1,161 @@
+#'   accurate progress reporting ([update_task_button()]).
+#' @param width Any valid CSS unit (e.g., `width="100%"`).
+#' @param submit_key A character string indicating what keyboard event should
+#'   trigger the submit button. The default is `enter`, which will submit the
+#'   input when the user presses the Enter/Return key. The `enter+modifier`
+#'   option will submit the input when the user presses the Enter key while
+#'   holding down Ctrl/Cmd.
+#'
+#' @return A textarea input control that can be added to a UI definition.
+#'
+#' @seealso [update_submit_textarea()], [input_task_button()]
+#'
+#' @examplesIf rlang::is_interactive()
+#'
+#' ui <- page_fluid(
+#'   input_submit_textarea("text", placeholder = "Enter some input..."),
+#'   verbatimTextOutput("value")
+#' )
+#' server <- function(input, output) {
+#'   output$value <- renderText({
+#'     req(input$text)
+#'     Sys.sleep(2)
+#'     paste("You entered:", input$text)
+#'   })
+#' }
+#' shinyApp(ui, server)
+#'
+#' @section Server value:
+#' A character string of the text input. The default value is `""` even if
+#' `value` is provided. The value will only be set/updated when the user submits
+#' the input by pressing the Enter key or clicking the submit button.
+#'
+#' @export
+input_submit_textarea <- function(
+  id,
+  label = NULL,
+  placeholder = NULL,
+  value = "",
+  ...,
+  button = NULL,
+  width = "min(680px, 100%)",
+  submit_key = c("enter", "enter+modifier")
+) {
+  rlang::check_installed("shiny", version = "1.11.1")
+  rlang::check_dots_empty()
+
+  value <- shiny::restoreInput(id = id, default = value)
+  if (length(value) != 1 || !is.character(value)) {
+    stop("`value` must be a character string", call. = FALSE)
+  }
+
+  submit_key <- rlang::arg_match(submit_key)
+  needs_modifier <- isTRUE(submit_key == "enter+modifier")
+
+  if (is.null(button)) {
+    button <- input_task_button(
+      id = paste0(id, "_submit"),
+      class = "btn-sm",
+      label = "Submit \U23CE",
+      title = "Press Enter to Submit",
+      `aria-label` = "Press Enter to Submit",
+      # The label/title above are updated in JS if a modifier key is needed
+      # (since it's platform dependent)
+      `data-default-button` = ""
+    )
+  }
+
+  if (!is_button_tag(button)) {
+    stop("`button` must be a `tags$button()`", call. = FALSE)
+  }
+
+  div(
+    class = "shiny-input-container bslib-mb-spacing",
+    style = css(
+      # TODO: validateCssUnit() needs to handle more complex CSS
+      width = if (is.numeric(width)) paste0(width, "px") else width,
+    ),
+    shiny_input_label(id, label),
+    div(
+      class = "bslib-input-textsubmit",
+      tags$textarea(
+        id = id,
+        class = "form-control",
+        style = css(width = if (!is.null(width)) "100%"),
+        placeholder = placeholder,
+        `data-needs-modifier` = if (needs_modifier) "",
+        rows = 1,
+        value
+      ),
+      button
+    )
+  )
```

Comment:
I'm curious if we could flatten this component one level, i.e. something like this

```
.bslib-input-textsubmit .shiny-input-container
  label
  textarea
  button
```

If that's possible, it'd be my preference as it'd make the structure a little easier. I guess in particular I don't like that `.bslib-input-textsubmit` is the sort-of key identifying class, but it's not at the top of the component, so you'd need something like `.shiny-input-container:has(> .bslib-input-text-submit)` to be able to find the component "root"